File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.js
function solution(n) {
    const dx = [0, 1, 0, -1];
    const dy = [1, 0, -1, 0];
    
    const arr = Array.from({ length: n }, () => Array(n).fill(0));
    
    let x = 0, y = 0, dir = 0;
    for (let i = 1; i <= n * n; i++) {
        arr[x][y] = i;

        let nx = x + dx[dir];
        let ny = y + dy[dir];

        if (nx < 0 || nx >= n || ny < 0 || ny >= n || arr[nx][ny] !== 0) {
            dir = (dir + 1) % 4;
            nx = x + dx[dir];
            ny = y + dy[dir];
        }

        x = nx;
        y = ny;
    }

    return arr;
}

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {
    public int[][] solution(int n) {
        int[][] arr = new int[n][n];
        
        int[] dx = new int[]{0, 1, 0, -1};
        int[] dy = new int[]{1, 0, -1, 0};
        
        int x = 0;
        int y = 0;
        int dir = 0;
        
        for (int i = 1; i <= n * n; i++) {
            arr[x][y] = i;
            
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            
            if (nx >= n || ny >= n || nx < 0 || ny < 0 || arr[nx][ny] != 0) {
                dir = (dir + 1) % 4;
                nx = x + dx[dir];
                ny = y + dy[dir];
            }
            
            x = nx;
            y = ny;
        }
        
        return arr;
    }
}